ARROW-14306: [C++][Compute] Add binary reverse kernel

johanpel opened the pull request.

lidavidm reviewed (approved):
Thanks!

Can we also add this function to https://github.com/apache/arrow/blob/master/docs/source/cpp/compute.rst ? (It should be in the same table as 'ascii_reverse' and 'utf8_reverse'.) We should document that this one does not make any assumptions about the encoding.

lidavidm commented:
Oh, and it needs to be listed in https://github.com/apache/arrow/blob/master/docs/source/python/api/compute.rst as well (again alongside its ascii/utf8 counterparts)

johanpel commented:
Thanks for the comments. I added the docs. I also added a test because the function can be applied to String/LargeString as well. I'm in doubt whether we would want to prevent applying this function to String/LargeString. Use cases would be quite exotic.

If we'd like to prevent it I would have to introduce `BinaryTypes()` with `{binary(), large_binary()}` only.

lidavidm commented:
Oh, whoops. No, good point: we should not support string types here, we should register only Binary and LargeBinary. For String/LargeString the kernel would need to validate that its output is still valid UTF-8 which would be redundant with string_reverse.

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -851,6 +851,8 @@ String transforms
 +-------------------------+--------+-----------------------------------------+------------------------+-----------------------------------+-------+
 | binary_replace_slice    | Unary  | String-like                             | Binary- or String-like | :struct:`ReplaceSliceOptions`     | \(5)  |
 +-------------------------+--------+-----------------------------------------+------------------------+-----------------------------------+-------+
+| binary_reverse          | Unary  | Binary                                  | Binary                 |                                   | \(11) |
```
Hmm... can you renumber the notes so that they appear in ascending order in this table? It's a bit cumbersome, but doable :-)

pitrou commented on docs/source/cpp/compute.rst:
```diff
@@ -911,6 +913,8 @@ String transforms
   If the input is not valid UTF8, then the output is undefined (but the size of output
   buffers will be preserved).
 
+* \(11) Performs a byte-level reverse. Does not make any assumptions about encoding.
```
```suggestion
* \(11) Perform a byte-level reverse.
```

pitrou reviewed (changes requested):
Thank you, just two nits.

lidavidm reviewed (approved):
Thanks for the update!
